internal.nix

```diff
@@ -33,6 +33,26 @@ rec {
   makeValidDrvName = str:
     lib.stringAsChars (c: if isValidDrvNameChar c then c else "?") str;
 
+  # Description: Checks if a string looks like a valid git revision
+  # Type: String -> Boolean
+  isGitRev = str:
+    (builtins.match "[0-9a-f]{40}" str) != null;
+
+  # Description: Takes a string of the format "git+http(s)://domain.tld/repo#commitish" and returns
+  # an attribute set { url, commitish }
+  # Type: String -> Set
+  parseGitRef = str:
+    let
+      parts = builtins.split "[#]" str;
+    in
+    assert !(builtins.length parts == 3) ->
+      builtins.throw "[npmlock2nix] failed to parse Git reference `${str}`. Expected a string of format `git+http(s)://domain.tld/repo#branch`";
+    rec {
+      inherit parts;
+      url = builtins.replaceStrings [ "git+" ] [ "" ] (builtins.elemAt parts 0);
```

Comment:
I am not liking this a lot. While it is unlikely that the `git+` string appears anywhere else in the URL this would replace parts of the URL that we shouldn't touch at all.

If we were to change this function to use `builtins.match` instead like so:
```
nix-repl> builtins.match "git\\+(https://[^#]+)#(.*)" "git+https://domain.tld/repo#commitish"
[ "https://domain.tld/repo" "commitish" ]
```

We extract exactly what we need from the URL (perhaps match the protocol explicitly?).


Untested proposal:
```nix
{ # ....
parseGitRef = let
     expression =  "git\\+(https://[^#]+)#(.*)"; # FIXME: whats the minimal commitish length? Enforce it here. Should be >= 1.
in str:
   assert builtins.typeOf str != "string" -> throw "parseGitRef expects a string that matches `${expression}`. Got `${builtins.typeOf str}` with value `${toString str}`";
  let
     m = builtins.match str;
   in
   assert m == null -> throw "parseGitRef must be called with a string of the format git+protocol://domain.tld/path#commitish but was called with `${str}`";
   assert builtins.length != 2 -> throw "parseGitRef got a URL that didn't conform to our expression `${expression}` with two captures groups. Matched as: `${toString m}`";
  {
    url = builtins.elemAt m 0;
    commitish = builtins.elemAt m 1;
  }
;
}
```